domain/state.go

```diff
@@ -119,3 +198,38 @@ func (r *txnRunner) Txn(ctx context.Context, fn func(context.Context, *sqlair.TX
 func (r *txnRunner) StdTxn(ctx context.Context, fn func(context.Context, *sql.Tx) error) error {
 	return CoerceError(r.runner.StdTxn(ctx, fn))
 }
+
+// TxContext is a typed context that provides access to the database transaction
+// for the duration of a transaction.
+type TxContext interface {
+	context.Context
+}
+
+// txContext is the concrete implementation of the TxContext interface.
+// The txContext ensures that a transaction is always available to during
+// the execution of a transaction.
+// The txContext stores the sqlair.TX directly on the struct to prevent the
+// need to fork the context during the transaction.
+// The mutex prevents data-races when the transaction is removed from the
+// context.
+type txContext struct {
+	context.Context
+
+	mu sync.Mutex
+	t  *sqlair.TX
+}
+
+// tx returns the sqlair.TX from the context.
+func (c *txContext) tx() *sqlair.TX {
+	c.mu.Lock()
```

Comment:
Can we ever really have concurrent access to `t`? 

At the point that you're in `RunTx` (as it's called now) all actions are locally scoped and synchronous.

Reply:
Not right now, as it only protects against a call to `Run` that happens-after the return of the `RunTx` closure. 

`Run` needs to hold the mutex for the duration of the `Run` closure for it to prevent an `sqlair.Tx`  being available to a state operation.

Obviously this only protects against violations on the service side, not violations on the state side. Since the state method is given the `sqlair.Tx`, `txContext.mu` has no power past that.

So state methods need to behave.